integration/integration_test.go

```diff
@@ -686,10 +686,41 @@ func TestCache(t *testing.T) {
 	}
 }
 
+func TestWarmer(t *testing.T) {
+	populateVolumeCache(t.Logf, config.serviceAccount)
+
+	for dockerfile := range imageBuilder.TestWarmerDockerfiles {
+		t.Run("test_warmer_"+dockerfile, func(t *testing.T) {
+			t.Parallel()
+			args, ok := additionalKanikoFlagsMap[dockerfile]
+			imageRepo := config.imageRepo
+			if !ok {
+				args = []string{}
+			}
+
+			// Build the initial without warmer
+			if err := imageBuilder.buildWarmerImage(t.Logf, config, dockerfilesPath, dockerfile, 0, args, false); err != nil {
+				t.Fatalf("error building cached image for the first time: %v", err)
+			}
+
+			// Build the second with warmer
+			if err := imageBuilder.buildWarmerImage(t.Logf, config, dockerfilesPath, dockerfile, 1, args, true); err != nil {
+				t.Fatalf("error building cached image for the second time: %v", err)
+			}
+
+			// Make sure both images are the same
+			kanikoVersion0 := GetKanikoImage(imageRepo, "test_warmer_"+dockerfile) + strconv.Itoa(0)
+			kanikoVersion1 := GetKanikoImage(imageRepo, "test_warmer_"+dockerfile) + strconv.Itoa(1)
+
+			containerDiff(t, kanikoVersion0, kanikoVersion1)
+			layerDiff(t, kanikoVersion0, kanikoVersion1)
+			manifestDiff(t, kanikoVersion0, kanikoVersion1)
+		})
+	}
+}
+
 // Attempt to warm an image two times : first time should populate the cache, second time should find the image in the cache.
 func TestWarmerTwice(t *testing.T) {
-	_, ex, _, _ := runtime.Caller(0)
-	tmpDir := filepath.Dir(ex) + "/tmpCache"
```

Comment:
oci warmer suffers from a time-of-check time-of-use issue here. We check whether the ocilayout exists, and if not, we create a tmpdir, store the ocilayout in there and copy it over. I don't think that this is new, it's just that now we store the images as folders as opposed to files, the rename will not just silently overwrite the existing cache entry.

Reply:
I muted the issue here in the integration tests by giving each thread an individual directory to store the images in. But in reality we will write to the same cache directory. So in case there are two warmer processes working on the same folder, ie. a shared volume serving multiple runners, now this can lead to spurious issues.

Where previously those issues would be silently resolved by overwriting each other's output, now the slower warmer will fail hard.

To resolve this issue we should add some mutex lock that is required for write access to the cache directory. To be safe the existence of a warmer entry must be checked again after acquiring the mutex-lock, as another process might have written the image by now.